### PR TITLE
ISSUE-787 Push for Settings

### LINKED ISSUE #787: Push for Settings
**Original ticket:** https://github.com/linagora/tmail-backend/issues/630

**WHY**

Upon setting update, we need to push the change event to FCM

**HOW**

create SettingTypeName object (extend TypeName), push the event StateChangeEvent to the event bus after updating setting

**DoD**: pass contract test

Notes: need Settings/set method available.

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve #787 
TODO check someplace are seem to always push StateChangeEvent even no successful change e.g `FilterSetMethod`

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/JmapSettingsSetMethod.scala [hunk 1]
```diff
@@ -97,7 +104,25 @@ class SettingsSetRequestSetMethod @Inject()(val jmapSettingsRepository: JmapSett
       oldState <- evaluateOldState(updateResults, currentStateIfAllUpdatesFailed)
       newState <- evaluateNewState(updateResults, currentStateIfAllUpdatesFailed)
       response = createResponse(invocation.invocation, request, updateResults, oldState, newState)
-    } yield InvocationWithContext(response, invocation.processingContext)
+    } yield {
+      dispatchSettingsChangeEvent(mailboxSession.getUser, oldState, newState).subscribe()
+
+      InvocationWithContext(response, invocation.processingContext)
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/JmapSettingsSetMethod.scala:110]: Why not just flatMap?

I do not see why we need to be async here.

Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/JmapSettingsSetMethod.scala:110]: I thought about 2 points:
- sync makes the response slower (even dispatch event should be quick, but still slower)
- In the case RabbitMQ is down, then failure to dispatch an event to RabbitMQ forces the Settings/set to fail. Seems not that critical to me.

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/JmapSettingsSetMethod.scala:110]: > sync makes the response slower (even dispatch event should be quick, but still slower)

IMO not a concern we should tackle here but more at the event bus layer.

Async here means anyway unbounded memory queue. Not sure that's a great idea...

> In the case RabbitMQ is down, then failure to dispatch an event to RabbitMQ forces the Settings/set to fail. Seems not that critical to me.

Again an event bus concern.

Event bus is supposed to save failure to dispatch into deadletter so I think that this concern is even already handled.

**Review by vttranlina — COMMENTED** [review]
read it

**Review by hungphan227 — APPROVED** [review]

**Review by Arsnael — COMMENTED** [review]
Nothing to add

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Needs rebase though

**Pull request merged**